### PR TITLE
CONSOLE-2892: Allow dynamic plugins to proxy to services on the cluster

### PR DESCRIPTION
**Pull request by jhadvig**
Based on https://github.com/openshift/enhancements/pull/915

ccing @vojtechszocs 

/assign @spadgett @christianvogt

### CODE REVIEW

File: console/v1alpha1/0000_10_consoleplugin.crd.yaml [hunk 1]
```diff
@@ -40,11 +40,47 @@ spec:
               type: object
               required:
                 - displayName
+                - proxy
```
Comment (spadgett) [inline console/v1alpha1/0000_10_consoleplugin.crd.yaml:43]: `proxy` shouldn't be required

File: console/v1alpha1/types_console_plugin.go [hunk 2]
```diff
@@ -27,16 +27,49 @@ type ConsolePluginSpec struct {
...
 	Service ConsolePluginService `json:"service"`
+	// proxy is a list of services that the plugin needs to connect to.
+	Proxy ConsolePluginProxy `json:"proxy"`
+}
+
+// ConsolePluginProxy holds information on various service types
+// to which console's backend will proxy the plugin's requests.
+type ConsolePluginProxy struct {
+	// services is list of Kubernetes Service to which the plugin
+	// will connect to, using a deployment with an HTTP server.
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:38]:
```suggestion
	// services is a list of in-cluster Services that the plugin
	// will connect to.
```

File: console/v1alpha1/types_console_plugin.go [hunk 3]
```diff
@@ -27,16 +27,49 @@ type ConsolePluginSpec struct {
...
+	// The Service must use HTTPS and service serving certificate.
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:39]: A service serving certificate isn't required. It's simply the default.

File: console/v1alpha1/types_console_plugin.go [hunk 4]
```diff
@@ -27,16 +27,49 @@ type ConsolePluginSpec struct {
...
+	// The console backend will proxy the communication between
+	// plugin and the specified Service.
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:41]: We should document the path that plugins can use to make requests as described in the enhancement.

File: console/v1alpha1/0000_10_consoleplugin.crd.yaml [hunk 5]
```diff
@@ -40,11 +40,47 @@ spec:
...
                 - service
               properties:
                 displayName:
                   description: displayName is the display name of the plugin.
                   type: string
+                proxy:
+                  description: proxy is a list of services that the plugin needs to connect to.
+                  type: object
+                  required:
+                    - services
```
Comment (spadgett) [inline console/v1alpha1/0000_10_consoleplugin.crd.yaml:53]: `services` shouldn't be required

File: console/v1alpha1/types_console_plugin.go [hunk 6]
```diff
@@ -27,16 +27,49 @@ type ConsolePluginSpec struct {
...
+	Services []ConsolePluginProxyService `json:"services"`
+}
+
+// ConsolePluginProxyService holds information on service to which
+// console's backend will proxy the plugin's requests.
+type ConsolePluginProxyService struct {
+	// name of Service that the plugin needs to connect to.
+	Name string `json:"name"`
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:49]: Let's make sure we have validation for these fields.

File: console/v1alpha1/types_console_plugin.go [hunk 7]
```diff
@@ -27,16 +27,49 @@ type ConsolePluginSpec struct {
...
+	// port on which the Service that the plugin needs to connect to
+	// is listening on.
+	Port int32 `json:"port"`
+	// caCertificate provides the cert authority certificate contents,
+	// in case the proxied service is using custom service CA.
+	// By default service CA bundle is used.
+	CACertificate string `json:"caCertificate"`
+	// authorize indicates if the proxied request will contain
+	// user's OpenShift access token. By default the access token
+	// is not part of the proxied request.
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:61]: You should mention how the token is passed in the `Authorization` request header.

**Comment by jhadvig** [comment]
PR is ready for final review.

QE Approver:
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

Console Approver:
/assign @spadgett

**Comment by jhadvig** [comment]
@spadgett I've addressed your comments. lets the `service CA` with lowercased starting 's' since kubernetes docs are using it this way as well.
PTAL

**Comment by yapei** [comment]
verified the changes in console-operator PR and it works 
/label qe-approved

File: console/v1alpha1/types_console_plugin.go [hunk 8]
```diff
@@ -15,32 +15,120 @@ type ConsolePlugin struct {
...
+	//
+	// +kubebuilder:validation:Optional
+	// +optional
+	Services []ConsolePluginProxyService `json:"services"`
+}
+
+// ConsolePluginProxyService holds information on Service to which
+// console's backend will proxy the plugin's requests.
+type ConsolePluginProxyService struct {
+	// name of Service that the plugin needs to connect to.
```
Comment (spadgett) [inline console/v1alpha1/types_console_plugin.go:65]: nit: we're using inconsistent punctuation here and below where we don't have a `.`

**Review by spadgett — APPROVED** [review]
/lgtm

**Pull request merged**